lib/madness/browser.rb

```diff
@@ -0,0 +1,63 @@
+require 'socket'
+require 'os'
+
+module Madness
+  # Handles browser launching
+  class Browser
+    attr_reader :host, :port
+
+    def initialize(host, port)
+      @host, @port = host, port
+    end
+
+    # Returns a URL based on host, port nad MADNESS_FORCE_SSL.
+    def server_url
+      scheme = ENV['MADNESS_FORCE_SSL'] ? 'https' : 'http'
+      url_host = ['0.0.0.0', '127.0.0.1'].include?(host) ? 'localhost' : host
+      "#{scheme}://#{url_host}:#{port}"
+    end
+
+    # Returns true if the server is running. Will attempt to connect
+    # multiple times. This is designed to assist in running some code after
+    # the server has launched.
+    def server_running?(retries: 5, delay: 1)
+      connected = false
+      attempts = 0
+
+      begin
+        connected = Socket.tcp(host, port)
+      rescue
+        sleep delay
+        retry if (attempts += 1) < retries
+      ensure
+        connected.close if connected
+      end
+
+      !!connected
+    end
+
+    # Open a web browser if the server is running. This is done in a
+    # non-blocking manner, so it can be executed before starting the server.
+    # It will yield an error message if it fails, or nil on success.
+    def open
+      fork do
+        if server_running?
+          success = open!
+          yield success ? nil : "Failed opening browser (#{open_command.join ' '})"
+        else
+          yield "Failed connecting to #{server_url}. Is the server running?"
+        end
+      end
+    end
+
```

Comment:
In general you are right, but in this case:

1. For now, we are just adding an "open browser" functionality, so dividing it to too many classes / modules, seems excessive. If future features will find their way to the `Browser` class area, then it will be a good time to consider refactoring some methods out of it.
2. Since the main API of the class is using `fork`, if the other methods are not public, they cannot be tested and testing will become much harder, since you will need to test all the cases through the fork.
3. Since this is a CLI and not a library, I do not mind habing some methods that should be private, made public for testability reasons. Besides, it is arguable that these should be private at all - someone might want to get the `#server_url`.
4. Once you see the specs, I believe it will become clearer why they are public.

Reply:
Yes sorry I was misunderstood. If testability comes at the price of them being public that is a good price! That is why I mentioned that I do not know too much about testing and for me private methods are somewhat ugly to test. But then I though: that if you know all the tricks and testing private methods is a thing you can and want to do, these methods might be private. Maybe I braindumped too much and reviewed too few, sorry about that.

Reply:
Heh, that's fine. I like philosophical Ruby discussions.

In my experience, testing should not use too many "tricks". It should be simple, easy to write and easy to read. Private methods should not be tested under any circumstances, since private methods mean that they are only accessed from other methods in the instance, which means the tests of the public methods should reflect all states, and all these states must reach all the corners of the private methods.